src/java_bytecode/java_types.h

```diff
@@ -18,6 +18,46 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/c_types.h>
 #include <util/optional.h>
 
+class java_annotationt : public irept
+{
+public:
+  class valuet : public irept
+  {
+  public:
+    valuet(irep_idt name, const exprt &value)
+      : irept(name)
+    {
+      get_sub().push_back(value);
+    }
+
+    const irep_idt &get_name()
+    {
+      return id();
+    }
+
+    const exprt &get_value()
+    {
+      return static_cast<const exprt &>(get_sub()[0]);
```

Comment:
I'm not a fan of magic constants. That includes magic index access.

Reply:
Please suggest alternative code.

Reply:
@LAJW Would `.front()` be more adequate?

Reply:
Nope. I'd like an explanation of what the first sub is. Add documentation at the top of the class specifying which ID does what; is there only one; etc. (there's none ATM).

But this is the problem with `irept`-based classes in general. There are no member variable names, only indices. There's no documentation for indices. In regular classes, even if there's no documentation, you can look at the header and see all members and their types at a glance. Here, the best we can do is just add documentation.

Also, an idea. Use constants named after the intended type: `static constexpr size_t FIELD_NAME=0, ...`. That way at least we'd get variable names.

Reply:
He wants this:
```
      get_sub() = irept::subt(value_id + 1);
      get_sub()[value_id] = value;
``
Unfortunately this is not a good idea.

Reply:
In this warped world of `irept` it is a goo idea.

Reply:
Example:

```c++

Reply:
Frankly horrifying inline data tables:

static boolean[] solid = {false, true, true, true, true, true, true, true, true, true

Reply:
But I don't care all that much, since I'm not the author, hence my approval.

Reply:
It might be redundant, but back to basics: there are multiple options for implementing expression trees. One option is to have a class for each kind of expression, with methods and fields specific to each kind of class. Another option is to have one very generic underlying storage and just put an API on top of that for each kind of expression. The latter provides less insight to the compiler (and thus less type checking), but is easier to extend. The former requires dedicated serialisation, deserialisation, hashing, etc. With `irept` we've got the latter. The ESBMC folks have by now moved over from `irept` to dedicated, fixed classes. You might check with @lucasccordeiro on the advantages and disadvantages of doing so.

For as long as we stick with `irept`, the purpose of a class such as the above is to provide a clear API and hide all implementation details (such as the placement of members in the internal storage). While and index `0` might be magic, it is also an information that is very local. The full scope of where one needs to know this fits on a single screen. Have we always got it perfectly right? No, there have been bugs (I do recall c33266a6f, maybe there have been others). The idea of using an enum instead of a magic number is probably a good one, as long as it doesn't get in the way of keeping the code compact.